search/request.go

```diff
@@ -0,0 +1,179 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package search
+
+import (
+	"encoding/json"
+	"github.com/tigrisdata/tigris-client-go/driver"
+	"github.com/tigrisdata/tigris-client-go/fields"
+	"github.com/tigrisdata/tigris-client-go/filter"
+)
+
+// Request for search
+type Request struct {
+	Q            string
+	SearchFields []string
+	Filter       filter.Filter
+	Facet        *FacetQuery
+	ReadFields   fields.Read
+	Options      *RequestOptions
+}
+
+func NewRequestBuilder(q string) RequestBuilder {
+	return &requestBuilder{
+		q:            q,
+		searchFields: make(map[string]bool)}
+}
+
+type RequestBuilder interface {
+	WithSearchFields(fields ...string) RequestBuilder
+	WithFilter(filter.Filter) RequestBuilder
+	WithFacet(*FacetQuery) RequestBuilder
+	WithReadOptions(fields.Read) RequestBuilder
+	WithOptions(*RequestOptions) RequestBuilder
+	Build() *Request
+}
+
+type requestBuilder struct {
+	q            string
+	searchFields map[string]bool
+	filter       filter.Filter
+	facet        *FacetQuery
+	readFields   fields.Read
+	options      *RequestOptions
+}
+
+func (b *requestBuilder) WithSearchFields(fields ...string) RequestBuilder {
+	for _, f := range fields {
+		b.searchFields[f] = true
+	}
+	return b
+}
+
+func (b *requestBuilder) WithFilter(f filter.Filter) RequestBuilder {
+	b.filter = f
+	return b
+}
+
+func (b *requestBuilder) WithFacet(facet *FacetQuery) RequestBuilder {
+	b.facet = facet
+	return b
+}
+
+func (b *requestBuilder) WithReadOptions(read fields.Read) RequestBuilder {
+	b.readFields = read
+	return b
+}
+
+func (b *requestBuilder) WithOptions(options *RequestOptions) RequestBuilder {
+	b.options = options
+	return b
+}
+
+func (b *requestBuilder) Build() *Request {
+	searchFields := make([]string, len(b.searchFields))
+	i := 0
+	for f := range b.searchFields {
+		searchFields[i] = f
+		i++
+	}
+
+	// default options
+	if b.options == nil {
+		b.options = &DefaultRequestOptions
+	}
+
+	return &Request{
+		Q:            b.q,
+		SearchFields: searchFields,
+		Filter:       b.filter,
+		Facet:        b.facet,
+		ReadFields:   b.readFields,
+		Options:      b.options,
+	}
+}
+
+type RequestOptions struct {
+	Page    int32
+	PerPage int32
+}
+
+var DefaultRequestOptions = RequestOptions{Page: 1, PerPage: 20}
+
+type FacetQuery struct {
+	FacetFields map[string]FacetQueryOptions
+}
+
+// Built marshals the facet query
+func (f *FacetQuery) Built() (driver.Facet, error) {
+	if f.FacetFields == nil || len(f.FacetFields) == 0 {
+		return driver.Facet(`{}`), nil
```

Comment:
if it is not set we don't need to send `{}` just send nil?

Reply:
read fields, filter are all being sent as `{}` to the server, hence I kept it to be consistent

Reply:
So when filters are sent as `{}` it means full table scan, for the read fields we don't need to but it in the same way it means right now read everything. The same doesn't apply to facets.

Reply:
Empty filters imply no filters to apply. I feel same could be the case for facets - no fields to facet for.

Read fields could also imply the same, but for user's convenience, we explicitly indicate that all fields will be read (basically no such projection to be applied).

I'll change it to nil, and on server we should treat `nil` and `{}` to be same. Faceting could be an expensive operation, on server we would have to cautiously evaluate *faceting on all fields*

Reply:
> Empty filters imply no filters to apply. I feel same could be the case for facets - no fields to facet for.

So there is a subtle difference here. Nil filter means an error, explicitly setting it to empty `{}` means doing a full table scan. This is an intentional behavior because this means we always need a filter object and it is not optional as per the API specs. On the other hand, this doesn't apply to facets. There is no concept of reading all facets and nil on facet simply means the user doesn't need facet query which is not the same as filters. 

> Read fields could also imply the same, but for user's convenience, we explicitly indicate that all fields will be read (basically no such projection to be applied)
Again the same thing, either read all which is the default behavior, or specify any specific field to read. But no option to not read anything. `{}` is used for Readfields just to be synonyms with FilterAll but with reads, even nil on the server-side is treated as read all and there is no meaning of `{}` on the server-side for read fields. 

> I'll change it to nil, and on server we should treat nil and {} to be same. Faceting could be an expensive operation, on server we would have to cautiously evaluate faceting on all fields
We don't need to do anything on the server, just treat it as nil which is optional.